File: api/my_private_finances/schemas/net_worth.py
from __future__ import annotations

from datetime import date
from decimal import Decimal

from my_private_finances.schemas.base import StrictSchema


class AccountBalancePoint(StrictSchema):
    account_id: int
    balance: Decimal


class NetWorthPoint(StrictSchema):
    month: str  # "YYYY-MM"
    total: Decimal
    by_account: list[AccountBalancePoint]


class AccountNetWorthSummary(StrictSchema):
    account_id: int
    account_name: str
    currency: str
    opening_balance: Decimal
    opening_balance_date: date
    current_balance: Decimal
    month_over_month_change: Decimal


class NetWorthReport(StrictSchema):
    currency: str
    current_total: Decimal
    month_over_month_change: Decimal
    accounts: list[AccountNetWorthSummary]
    history: list[NetWorthPoint]  # monthly data points, oldest first

File: app/src/lib/api/netWorth.ts
import { apiGet } from "./client";

export type AccountBalancePoint = {
  account_id: number;
  balance: string;
};

export type NetWorthPoint = {
  month: string; // "YYYY-MM"
  total: string;
  by_account: AccountBalancePoint[];
};

export type AccountNetWorthSummary = {
  account_id: number;
  account_name: string;
  currency: string;
  opening_balance: string;
  opening_balance_date: string;
  current_balance: string;
  month_over_month_change: string;
};

export type NetWorthReport = {
  currency: string;
  current_total: string;
  month_over_month_change: string;
  accounts: AccountNetWorthSummary[];
  history: NetWorthPoint[];
};

export function getNetWorth(months: number = 12): Promise<NetWorthReport> {
  return apiGet<NetWorthReport>(`/api/reports/net-worth?months=${months}`);
}
